### PR TITLE
Add ExtractPublicKey function

### PR DESCRIPTION
**Pull request by JustinDrake**
This is a draft function to extract the public key from the identity. It is a helper function for https://github.com/ipfs/go-ipfs/issues/3896 which will be used in `go-libp2p-kad-dht` and `go-ipfs/namesys`. I'll add tests when the IPFS team is happy with the approach.

A few questions:

1. Should these helper functions live in `go-libp2p-peer`?
2. I designed the functions to return the nil public key if key extraction fails. Is that acceptable?
3. I've added an `UnmarshalEd25519PublicKey` function (see [PR here](https://github.com/libp2p/go-libp2p-crypto/pull/24/files)). Is that OK?

cc @whyrusleeping, @Kubuxu

### CODE REVIEW

**Review by ghost — CHANGES_REQUESTED** [review]
This largely LGTM and I think go-libp2p-peer is the right place, but I feel slightly uncomfortable throwing away all errors. There should be at least a "wasn't able to extract pubkey" error returned.

Also needs tests for the happy path and for error cases.

File: peer.go [hunk 1]
```diff
@@ -63,6 +64,64 @@ func (id ID) MatchesPublicKey(pk ic.PubKey) bool {
 	return oid == id
 }
 
+func (id ID) ExtractEd25519PublicKey() ic.PubKey {
+	// ed25519 pubkey identity format
+	// <identity mc><length (2 + 32 = 34)><ed25519-pub mc><ed25519 pubkey>
+	// <0x00       ><0x22                ><0xed01        ><ed25519 pubkey>
```
Comment (Kubuxu) [inline peer.go:70]: I am not sure if `0xed01` is the varint encoded multicodec of the ed25519-pub code

Comment (Kubuxu) [reply peer.go:70]: Ahh, the check is correct the doc is just off.

File: peer.go [hunk 2]
```diff
@@ -63,6 +64,64 @@ func (id ID) MatchesPublicKey(pk ic.PubKey) bool {
...
+		return nilPubKey
+	}
+
+	// Check ID multihash codec
+	if decoded.Code != mh.ID {
+		return nilPubKey
+	}
+
+	// Check multihash length
+	if decoded.Length != 2+32 {
```
Comment (Kubuxu) [inline peer.go:86]: And this might be off.

File: peer.go (hunk 1 above)
Comment (JustinDrake) [reply peer.go:70]: Is there a Go function that will varint encode a byte slice for me?

The ed25519-pub code `0xed` is `0b11101101` in binary. The last byte needs to have it's MSB set to 0 (according to [this](https://richardschneider.github.io/net-ipfs-core/api/Ipfs.Varint.html)) and each byte except the last contains 7 bits of information. So I imagine the MSB (which is 1) gets chopped off from the first byte and added to the next byte. That would yield `0b(1)1101101(0)1000000`. Or is it that we shift everything by one bit, yielding `0b(1)1110110(0)1000000`? I'm so confused by varint!

File: peer.go (hunk 2 above)
Comment (JustinDrake) [reply peer.go:86]: Why is this off? I think the varint-encoded ed25519-pub code is 2 bytes and the ed25519 pubkey is 32 bytes.

**Comment by JustinDrake** [comment]
@lgierth 

> I feel slightly uncomfortable throwing away all errors

The way I see it is that the default case for the functions is to fail to extract a public key. If we have `n` identity types then the extraction function will fail for the other `n-1` identity types. Hence in this context "failure" is not really a failure.

Furthermore the error in `pubKey, err := ic.UnmarshalEd25519PublicKey(pubKeyBytes)` should never happen. The reason is that the length check in `UnmarshalEd25519PublicKey` will never fail because the `if decoded.Length != 2+32` length check comes before. I just added the error check here as a formality.

As for the error in `decoded, err := mh.Decode([]byte(id))` any such error should be caught upstream as it is a very low-level error.

File: peer.go (hunk 1 above)
Comment (Kubuxu) [reply peer.go:70]: You can do binary.PutUvarint with the value. It is compliant with our varint spec up until 63 bits of to be encoded value.

Comment (JustinDrake) [reply peer.go:70]: Ok. It seems the documentation is correct with `0xed01` (see [code here](https://play.golang.org/p/12Lw8AOn1l)). Could you confirm this? And could you expand on why you think that the length might be off in [this comment](https://github.com/libp2p/go-libp2p-peer/pull/13#discussion_r116257739).

File: peer.go (hunk 2 above)
Comment (Kubuxu) [reply peer.go:86]: Sorry I think it should be ok, I accounted for length prefix.

File: peer.go [hunk 3]
```diff
@@ -63,6 +64,64 @@ func (id ID) MatchesPublicKey(pk ic.PubKey) bool {
...
+func (id ID) ExtractEd25519PublicKey() ic.PubKey {
+	// ed25519 pubkey identity format
+	// <identity mc><length (2 + 32 = 34)><ed25519-pub mc><ed25519 pubkey>
+	// <0x00       ><0x22                ><0xed01        ><ed25519 pubkey>
+
+	var nilPubKey ic.PubKey
+
+	// Decode multihash
+	decoded, err := mh.Decode([]byte(id))
+	if err != nil {
```
Comment (whyrusleeping) [inline peer.go:76]: Hrm... maybe we log an error or panic or something in this case? This is essentially an invariant of the type, no?

cc @Kubuxu @lgierth

**Comment by whyrusleeping** [comment]
@JustinDrake similar thoughts as @lgierth on the error handling, we shouldnt throw them away, lets log them at the very least.

Also, do we need both methods? Or can we just have the generic `ExtractPubKey` and then the caller can check the type on their own? I hesitate adding too much extra complexity to this type

**Comment by CMCDragonkai** [comment]
Is the extraction of public keys from peer ids a necessity of the ipfs infrastructure, or is it only used for showing the user the public keys for reporting/logging/user interface? Because not all peer ids allow the public key to be extracted. And in fact the current code base only shows that Ed25519 peer ids allow the public key to be extracted. According to various descriptions of the peer id it appears that in general the public key shouldn't be extractable.

Would it be better such that public keys can always be extracted from a peer id (hence you're just using the public key as the peer id)?

In terms porting this to Haskell, we would like to know exactly when a conversion of PeerID -> PublicKey is actually meant to be allowed.

**Comment by JustinDrake** [comment]
In the general case keys are not extractable, because keys can be larger, and identities are meant to be short. Currently the only extractable case is with Ed25519, when the identity multihash is used, as well as the Ed25519-pub multicodec. See https://github.com/libp2p/go-libp2p-peer/blob/master/peer.go#L76

**Pull request closed without merging**